### PR TITLE
Factor out direct Tensor access into a serializable data structure

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121821

Context: https://github.com/pytorch/pytorch/issues/121085

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1824,6 +1824,29 @@ def print_results(grouped, indent, result_fn):
 TLS = threading.local()
 
 
+@dataclass(frozen=True)
+class CreateSymbolicSizesStridesStorageOffsetInput:
+    size: Tuple[int, ...]
+    stride: Tuple[int, ...]
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:1830]: `Union[int, Symbol]`, I guess

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -2277,12 +2300,23 @@ def _produce_dyn_sizes_from_int_tuple(self,
...
 
+    # TODO: This is just for testing convenienced, rename accordingly
     def create_symbolic_sizes_strides_storage_offset(
         self,
         ex: torch.Tensor,
         source: Source,
         *,
         symbolic_context: Optional[SymbolicContext] = None,
+    ):
+        inp = CreateSymbolicSizesStridesStorageOffsetInput.from_tensor(ex)
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:2311]: Java-inspired names lol

**Review by lezcano — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1824,6 +1824,29 @@ def print_results(grouped, indent, result_fn):
 TLS = threading.local()
 
 
+@dataclass(frozen=True)
+class CreateSymbolicSizesStridesStorageOffsetInput:
```
Comment (eellison) [inline torch/fx/experimental/symbolic_shapes.py:1828]: would it make sense to name this something more generic if we anticipate it encompassing things like Sparse tensors in the future ? (i dont know, asking)

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -1824,6 +1824,29 @@ def print_results(grouped, indent, result_fn):
...
+            ex.stride(),
+            ex.storage_offset(),
+            getattr(ex, "_dynamo_dynamic_indices", set())
+        )
+
+    def dim(self):
+        return len(self.size)
+
+    def is_dim_dynamic(self, i):
+        return i in self.dynamo_dynamic_indices
```
Comment (eellison) [inline torch/fx/experimental/symbolic_shapes.py:1847]: nit: type annotations.

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -2277,12 +2300,23 @@ def _produce_dyn_sizes_from_int_tuple(self,
...
         self,
         ex: torch.Tensor,
         source: Source,
         *,
         symbolic_context: Optional[SymbolicContext] = None,
+    ):
+        inp = CreateSymbolicSizesStridesStorageOffsetInput.from_tensor(ex)
+        return self.hermetic_create_symbolic_sizes_strides_storage_offset(inp, source, symbolic_context=symbolic_context)
+
+    def hermetic_create_symbolic_sizes_strides_storage_offset(
```
Comment (eellison) [inline torch/fx/experimental/symbolic_shapes.py:2314]: I find hermetic here to be a little bit confusing. would `create_symbolic_sizes_strides_storage_offset_from_tensor` method above renaming, and `create_symbolic_sizes_strides_storage_offset` naming here make more sense ?


There's not reallyyyy anything more hermetic from this method and the other. They both do the same thing using the ex input to get the required metadata

**Review by eellison — APPROVED** [review]

**Comment by ezyang** [comment]
obsoleted by https://github.com/pytorch/pytorch/pull/122044

**Pull request closed without merging**